torch/csrc/lazy/core/lazy_mode.cpp

```diff
@@ -0,0 +1,75 @@
+#include <torch/csrc/lazy/core/lazy_mode.h>
+
+#include <c10/core/DispatchKey.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+#include <torch/csrc/lazy/backend/backend_device.h>
+#include <torch/csrc/lazy/core/lazy_graph_executor.h>
+#include <torch/library.h>
+
+// TODO(whc) we can't actually depend on ts backend code from here, but we could refactor if reusing the callback
+// turns out to be the best way to implement this
+#include <torch/csrc/lazy/ts_backend/ts_eager_fallback.h>
+
+namespace torch {
+namespace lazy {
+
+std::atomic<bool>& in_lazy_mode() {
+    static std::atomic<bool> _in_lazy_mode{false};
+    return _in_lazy_mode;
+}
+
+void LazyModeEnter(c10::Device device) {
+    in_lazy_mode() = true;
+    // It is straightforward why we want to set the lazy key on entering the mode:
+    // we force operators (even on regular eager tensors) to route to lazy implementations
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Lazy, true);
+}
+
+void LazyModeExit(c10::Device device) {
+    // Equally straightforward is that we no longer want the lazy key when we exit the mode:
+    // this lets operations on eager tensors outside the mode go back to normal eager behavior
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Lazy, false);
+
+    // Less obvious is that we also set an 'unlazy' key on mode exit, which lets us specially handle
+    // any 'lazy' tensors that are alive after the mode exit.  This could be avoided if we can find another way
+    // to make lazy tensors interoperable with eager kernels.
+    // TODO(whc) PrivateUse1 is just for prototyping
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::PrivateUse1, true);
```

Comment:
IIUC `PrivateUse1` here is in place of `unlazy`?

https://github.com/pytorch/pytorch/blob/wconstab/lazymode/c10/core/DispatchKey.cpp#L73-L85
`Meta` does seem to have a higher priority than `PrivateUse1`, not sure if I'm answering your question in the failure cases....

Reply:
I thought it was the opposite: PrivateUse1 is higher priority than Meta.  (Keys lower in the list have higher enum values and run first).  This is true e.g. that Lazy runs before CPU/CUDA and Autograd* runs before Lazy.    

Actually, Meta should not even be a factor: the meta key should not be set on any of our tensors, since they are not meta tensors, so the priority of the Meta key should be no factor.

I just realized that there are 2 separate reasons for meta kernels running in the two separate tests.
- test_lazy_mode(): error happens on lazy add operation with inputs (lazy, cuda).  Lazy add actually calls the meta kernel, (which is expected) but it feeds it a lazy tensor and a cuda tensor.  This part needs to be fixed in the lazy tensor code
- test_single_operations(): fails outside of lazy mode, when the eager/cuda add happens on inputs (lazy, cuda).   It also calls the meta kernel directly, to check inputs.  However, this case isn't a direct bug, it should be fixed by the unlazy kernel running, which it is not.